(GH-2629) Add top-level metaparameter keys to YAML plan steps (#2677)
puppetlabs/bolt: documentation/writing_yaml_plans.md

## Plan structure

> YAML plans contain a list of steps and can include additional optional keys.

Reviewer: ```suggestion
YAML plans contain a list of steps and can include optional top-level keys.
```

> The following top-level keys are available:

Reviewer: These formatting improvements are SO MUCH BETTER! I feel like at some point we'll want to add a reference doc for YAML plan keys, but definitely doesn't have to be here (I'm also like 99% sure you've had the same idea :stuck_out_tongue_winking_eye:)

Author: You know me so well 😅

| Key | Type | Description | Required |
| --- | --- | --- | --- |
| `description` | `String` | The plan description. Appears in `bolt plan show <PLAN NAME>` and `Get-BoltPlan <PLAN NAME>` output. | |
| `parameters` | `Hash` | A hash of plan parameters. Each key is the name of the parameter and the value is the parameter definition. | |
| `private` | `Boolean` | Whether the plan should appear in `bolt plan show` and `Get-BoltPlan` output. | |
| `return` | `Array`, `Boolean`, `Hash`, `Number`, `String` | The value to return from the plan. Must evaluate to a valid [PlanResult](bolt_types_reference.md#planresult). | |
| `steps` | `Array` | The list of steps to run. | ✓ |

## Steps

The `steps` key is an array of step objects, each of which corresponds to a
specific action to take.

> When the plan runs, each step is executed in order. By default, if a step fails
> the plan halts execution and raises an error containing the result of the step
> that failed.

Reviewer 2: ```suggestion
When the plan runs, each step is executed in order. If a step fails,
the plan halts execution and raises an error containing the result of the step
that failed.
```
Do we need the "by default" here? It's a bit awkward with the conditional cos you need another comma.

### Message step

> Use a `message` step to print a message. This will print a message to standard
> out (stdout) when using the `human` output format, and print to standard error
> (stderr) when using the `json` output format. 

Reviewer 2: ```suggestion
Use a `message` step to print a message. The step prints a message to standard
out (stdout) when using the `human` output format, and prints to standard error
(stderr) when using the `json` output format. 
```

For information on printing a step result with `message`, see [Debugging
plans](#debugging-plans).

### Command step

Use a `command` step to run a single command on a list of targets and save the
results, containing stdout, stderr, and exit code. The step fails if the exit
code of any command is non-zero.

Command steps support the following keys:

| Key | Type | Description | Required |
| --- | --- | --- | --- |

> | `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |

Reviewer 2: ```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

    description: "Get the webserver hostnames"
```

### Task step

Use a `task` step to run a Bolt task on a list of targets and save the results.

Task steps support the following keys:

| Key | Type | Description | Required |
| --- | --- | --- | --- |

> | `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |

Reviewer 2: ```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

    description: "Upload motd to the webservers"
```

### Plan step

Use a `plan` step to run another plan and save its result.

Plan steps support the following keys:

| Key | Type | Description | Required |
| --- | --- | --- | --- |

> | `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |

Reviewer 2: ```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```

`apply_prep` see the [Applying manifest blocks](applying_manifest_blocks.md#)
section.

Resources steps support the following keys:

| Key | Type | Description | Required |
| --- | --- | --- | --- |

> | `catch_errors` | `Boolean` | Whether to catch raised errors. If set to true, the plan will continue execution if the step fails. | |

Reviewer 2: ```suggestion
| `catch_errors` | `Boolean` | Whether to catch raised errors. If set to `true`, the plan continues execution if the step fails. | |
```
